compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -221,6 +221,103 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
+                    let field_tys = variant.fields.iter().filter_map(|field| {
+                        let ty = tcx.type_of(field.did).instantiate(tcx, args);
+                        match (field.vis, module) {
+                            // The field is public or we do not care for visibility.
+                            (ty::Visibility::Public, _) | (_, None) => Ok(ty),
+                            (ty::Visibility::Restricted(from), Some(module)) => {
+                                if tcx.is_descendant_of(module, from) {
+                                    Ok(ty)
+                                } else {
+                                    // From `module`, this field may be privately uninhabited.
+                                    // Do not consider it as uninhabited.
+                                    Err(NoSolution)
+                                }
+                            }
+                        }
+                        .ok()
+                    });
+                    self.compute_uninhabited_goal_fields(param_env, module, field_tys)?;
+                }
+                self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+            }
+            ty::Array(ty, len) => match len.try_eval_target_usize(tcx, param_env) {
+                None | Some(0) => Err(NoSolution),
+                Some(1..) => {
+                    self.add_goal(Goal::new(
+                        tcx,
+                        param_env,
+                        ty::PredicateKind::Uninhabited(*ty, module),
+                    ));
+                    self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                }
+            },
+            ty::Alias(ty::Opaque, _) => Err(NoSolution),
+            // FIXME
+            ty::Alias(ty::Projection | ty::Inherent, _) => Err(NoSolution),
+            // use a query for more complex cases
+            // references and other types are inhabited
+            _ => Err(NoSolution),
+        }
+    }
+
+    #[instrument(level = "debug", skip(self, components))]
+    fn compute_uninhabited_goal_fields(
+        &mut self,
+        param_env: ty::ParamEnv<'tcx>,
+        module: Option<DefId>,
+        components: impl Iterator<Item = Ty<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        let responses: Vec<_> = components
+            .filter_map(|ty| {
+                self.probe_misc_candidate("uninhabitedness candidate")
+                    .enter(|ecx| {
+                        ecx.add_goal(Goal::new(
+                            ecx.tcx(),
+                            param_env,
+                            ty::PredicateKind::Uninhabited(ty, module),
+                        ));
+                        ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                    })
+                    .ok()
+            })
+            .collect();
+        self.try_merge_responses(&responses).ok_or(NoSolution)
```

Comment:
need to `flounder` if that returns `None`, we also return `None` if there are are multiple candidates which cannot be merged. In this case we want to return ambiguity instead of `NoSolution`